node/overseer/src/lib.rs

```diff
@@ -919,36 +952,40 @@ where
 	async fn block_imported(&mut self, block: BlockInfo) -> SubsystemResult<()> {
 		let mut update = ActiveLeavesUpdate::default();
 
-		if let Some(parent) = block.number.checked_sub(1).and_then(|number| self.active_leaves.take(&(block.parent_hash, number))) {
-			update.deactivated.push(parent.0);
-			self.metrics.on_head_deactivated();
+		if let Some(_number) = self.active_leaves.remove(&block.parent_hash) {
+			update.deactivated.push(block.parent_hash);
+			self.on_head_deactivated(&block.parent_hash);
 		}
 
-		if !self.active_leaves.contains(&(block.hash, block.number)) {
+		if self.active_leaves.get(&block.hash).is_none() {
 			update.activated.push(block.hash);
-			self.active_leaves.insert((block.hash, block.number));
-			self.metrics.on_head_activated();
+			self.active_leaves.insert(block.hash, block.number);
+			self.on_head_activated(&block.hash);
 		}
 
+		self.clean_up_external_listeners();
```

Comment:
the only downside to this is if a lot of hashes are interesting so this is slow, or if block importing stopped for some reason, but both scenarios seem unlikely

Reply:
Agree.